File: 29. Shortest Subarray with Sum at least K.cpp
// Topic: Shortest Subarray with Sum at least K
#include <bits/stdc++.h>
using namespace std;

int shortestSubarray(vector<int>& A, int K)
{
    int n = A.size();
    vector<long long> prefix(n + 1, 0);
    for (int i = 0; i < n; ++i) prefix[i + 1] = prefix[i] + A[i];
    deque<int> dq;
    int res = INT_MAX;
    for (int i = 0; i <= n; ++i)
    {
        while (!dq.empty() && prefix[i] - prefix[dq.front()] >= K)
        {
            res = min(res, i - dq.front());
            dq.pop_front();
        }
        while (!dq.empty() && prefix[i] <= prefix[dq.back()]) dq.pop_back();
        dq.push_back(i);
    }
    return res == INT_MAX ? -1 : res;
}

int main()
{
    vector<int> A = {2, -1, 2};
    int K = 3;
    cout << "Min length: " << shortestSubarray(A, K) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 29. Shortest Subarray with Sum at least K.py
# Topic: Shortest Subarray with Sum at least K (deque on prefix sums)

from collections import deque

def shortest_subarray(A, K):
    n = len(A)
    prefix = [0]*(n+1)
    for i in range(n): prefix[i+1] = prefix[i] + A[i]
    dq = deque()
    res = float('inf')
    for i in range(n+1):
        while dq and prefix[i] - prefix[dq[0]] >= K:
            res = min(res, i - dq.popleft())
        while dq and prefix[i] <= prefix[dq[-1]]:
            dq.pop()
        dq.append(i)
    return -1 if res == float('inf') else res

if __name__ == '__main__':
    A = [2,-1,2]
    K = 3
    print('Min length:', shortest_subarray(A,K))

# Time Complexity: O(n)
# Space Complexity: O(n)
